pkg/services/baremetal/baremetal/baremetal.go

```diff
@@ -504,6 +453,69 @@ func (s *Service) chooseHost(ctx context.Context) (
 	return chosenHost, helper, "", nil
 }
 
+func (s *Service) skipHost(labelSelector labels.Selector, host infrav1.HetznerBareMetalHost, mapOfSkipReasons map[string]int) bool {
+	// This comes first, because we should not look too deep into machines
+	// which are not in our scope.
+	if !labelSelector.Matches(labels.Set(host.ObjectMeta.Labels)) {
+		mapOfSkipReasons["label-selector-does-not-match"]++
+		return true
+	}
+
+	if host.GetDeletionTimestamp() != nil {
+		mapOfSkipReasons["hbmh-has-deletion-timestamp"]++
+		return true
+	}
+
+	if host.Spec.MaintenanceMode != nil && *host.Spec.MaintenanceMode {
+		mapOfSkipReasons["hbmh-in-maintenance-mode"]++
+		return true
+	}
+	if host.Spec.Status.ErrorMessage != "" {
+		mapOfSkipReasons["hbmh-has-error-message-in-status"]++
+		return true
+	}
+
+	if host.Spec.Status.ProvisioningState != infrav1.StateNone {
+		mapOfSkipReasons["hbmh-in-wrong-provisioning-state"]++
+		return true
+	}
+
+	if host.Spec.RootDeviceHints == nil ||
+		(host.Spec.RootDeviceHints.WWN == "" && len(host.Spec.RootDeviceHints.Raid.WWN) == 0) {
+		// Even if there are no rootDeviceHints specified, the host should be picked.
+		// After the phase registering, the process to provision the server stops and
+		// waits for the user to specify the rootDeviceHints.
+		// Here (RootDeviceHints exists and is valid) we want to check whether
+		// the specified rootDeviceHints fit with the InstallImage configuration
+		// of the HetznerBareMetalMachine. If not, it is not valid.
+		// Doing that without first choosing the hbmh would be nice, there is a feature request:
+		// https://github.com/syself/cluster-api-provider-hetzner/issues/1166
+		// See "tworaidchecks" for the other place.
+		return false
+	}
+
+	if !host.Spec.RootDeviceHints.IsValid() {
```

Comment:
you should add one small comment to make this understandable together with the previous check:

 // IsValid returns false if rootDeviceHints are empty, but this case was handled before